update-versions.py

```diff
@@ -1,50 +1,90 @@
 import argparse
 import collections
 import functools
-import github
 import json
 import os
 import pathlib
-import semver
 import subprocess
+from dataclasses import dataclass
+from pathlib import Path
+from typing import (
+    Callable,
+    Iterable,
+    List,
+    Optional,
+    OrderedDict,
+)
 
+import github
+import semver
+from github.GitRelease import GitRelease
 
-def read_versions():
-    with open("versions.json", "r") as f:
-        return json.load(f)
 
+@dataclass
+class NixHashes:
+    hash: str
+    vendorHash: str
 
-def is_stable(release):
-    version = release.tag_name.removeprefix("v")
-    return semver.compare(version, "1.0.0") >= 0 and not (
-        release.draft or release.prerelease
-    )
 
+Versions = OrderedDict[semver.Version, NixHashes]
+
+
+def parse_semver(input: str) -> semver.Version:
+    return semver.Version.parse(input.removeprefix("v"))
+
+
+def read_current_versions(file: Path) -> Versions:
+    with open(file, "r") as f:
+        versions = json.load(f)
+
+    result: OrderedDict[semver.Version, NixHashes] = OrderedDict()
+    for key_raw, value_raw in versions.items():
+        key = parse_semver(key_raw)
+        result[key] = NixHashes(
+            hash=value_raw["hash"], vendorHash=value_raw["vendorHash"]
+        )
+
+    return result
 
-def by_version(release):
-    return release.tag_name.removeprefix("v").split(".")
 
+def get_stable_github_versions(releases: Iterable[GitRelease]) -> List[semver.Version]:
+    stable_version = semver.Version(1, 0, 0)
 
-def to_version(vendor_hash):
-    def add_version(versions, release):
-        version = release.tag_name.removeprefix("v")
-        calculated_hash = calculate_hash(versions, version)
-        versions[version] = {
-            "hash": calculated_hash,
-            "vendorHash": calculate_vendor_hash(
-                versions, version, calculated_hash, vendor_hash
-            ),
-        }
-        return versions
+    result: List[semver.Version] = []
+    for release in releases:
```

Comment:
@pedorich-n I would argue that using high-order functions makes it a little bit easier to reason what this specific section of the code does as it hides some of the implementation details of looping over the elements of the list, filtering, and generating a new list out of the subset of stable releases:

```py
return map(to_version, filter(is_stable, releases))
```

While I understand the proposed approach involves looping through the list of releases only once, I don't think we have a lot of entries right now to worry about performance, so I'm more inclined to favor readability. What are your thoughts?

Reply:
As I primarily use FP languages, I agree, but I think in Python FP is not the preferred way of doing things. I've even heard the `map`, `filter`, `reduce`, and other functions were supposed to be removed from Python 3.
But if you don't mind that, I sure can change to `map` and `filter`

Reply:
Done